spec/models/transformation_mapping_item_spec.rb

```diff
@@ -45,4 +46,60 @@
       expect(invalid_mapping_item.errors[:destination].first).to match("EMS type of destination cluster or cloud tenant must be in")
     end
   end
+  # end of destination cluster validation
+
+  context "datastore validation" do
+    let(:src_host) { FactoryBot.create(:host_vmware, :ems_cluster => vmware_cluster) }
+    let(:src)      { FactoryBot.create(:storage_vmware, :hosts => [src_host]) }
+
+    let(:ems) { FactoryBot.create(:ems_openstack) }
+    let(:disk) { FactoryBot.create(:disk) }
+    let(:cloud_tenant) { FactoryBot.create(:cloud_tenant_openstack, :ext_management_system => ems) }
+    let(:cloud_volume_openstack) { FactoryBot.create(:cloud_volume_openstack, :attachments => [disk], :cloud_tenant => cloud_tenant) }
+
+    before do
+      allow(vmware_cluster).to receive(:storages).and_return([src])
+    end
+
+    context "source validation" do
+      let(:valid_tmi_vmw) { FactoryBot.create(:transformation_mapping_item, :source => src, :destination => cloud_volume_openstack) }
+      let(:invalid_tmi_vmw) { FactoryBot.build(:transformation_mapping_item, :source => cloud_volume_openstack, :destination => cloud_volume_openstack) }
+
+      it "Source datasource is valid" do
+        expect(valid_tmi_vmw.valid?).to be(true)
+      end
+
+      it "Source datasource is invalid" do
+        expect(invalid_tmi_vmw.valid?).to be(false)
+      end
+    end
+    # end of vmware source
+
+    context "destination validation" do
+      context "valid destination" do
+        let(:valid_tmi_ops) { FactoryBot.create(:transformation_mapping_item, :source => src, :destination => cloud_volume_openstack) }
+
+        it "openstack is valid" do
+          expect(valid_tmi_ops.valid?).to be(true)
+        end
+      end
+      # end of openstack destination
+
+      context "redhat" do
+        let(:dst_host_rh) { FactoryBot.create(:host_redhat, :ems_cluster => redhat_cluster) }
+        let(:dst_rh)      { FactoryBot.create(:storage_nfs, :hosts => [dst_host_rh]) }
+        before do
+          allow(redhat_cluster).to receive(:storages).and_return([dst_rh])
```

Comment:
Same here

Reply:
This spec is going to change because the way I filter now has changed.